deepspeed/runtime/bf16_optimizer.py

```diff
@@ -0,0 +1,582 @@
+                             allgather_bucket_size=self.allgather_bucket_size)
+
+        self.clear_hp_grads()
+        self.step_count += 1
+
+    def backward(self, loss, update_hp_grads=True, clear_lp_grads=False, **bwd_kwargs):
+        """Perform a backward pass and copy the low-precision gradients to the
+        high-precision copy.
+
+        We copy/accumulate to the high-precision grads now to prevent accumulating in the
+        bf16 grads after successive backward() calls (i.e., grad accumulation steps > 1)
+
+        The low-precision grads are deallocated during this procedure.
+        """
+        self.clear_lp_grads()
+        loss.backward(**bwd_kwargs)
+
+        if update_hp_grads:
+            self.update_hp_grads(clear_lp_grads=clear_lp_grads)
+
+    @torch.no_grad()
+    def update_hp_grads(self, clear_lp_grads=False):
+        for i, group in enumerate(self.bf16_groups):
+            for j, lp in enumerate(group):
+                if lp.grad is None:
+                    continue
+
+                hp_grad = self.fp32_groups_gradients[i][j]
+                assert hp_grad is not None, \
+                    f'high precision param has no gradient, lp param_id = {id(lp)} group_info = [{i}][{j}]'
+
+                hp_grad.data.add_(lp.grad.data.to(hp_grad.dtype).view(hp_grad.shape))
+                lp._hp_grad = hp_grad
+                self.fp32_groups_has_gradients[i][j] = True
+
+                # clear gradients
+                if clear_lp_grads:
+                    lp.grad = None
+
+    @torch.no_grad()
+    def get_grads_for_reduction(self):
+        return self.fp32_groups_gradients_flat
+
+    @torch.no_grad()
+    def get_grads_for_norm(self, for_clipping=False):
+        grads = []
+        tensor_mp_rank = bwc_tensor_model_parallel_rank(mpu=self.mpu)
+        for i, group in enumerate(self.bf16_groups):
+            for j, lp in enumerate(group):
+                if not for_clipping:
+                    if hasattr(lp, PIPE_REPLICATED) and lp.ds_pipe_replicated:
+                        continue
+
+                    if not (tensor_mp_rank == 0 or is_model_parallel_parameter(lp)):
+                        continue
+
+                if not self.fp32_groups_has_gradients[i][j]:
+                    continue
+
+                grads.append(self.fp32_groups_gradients[i][j])
+
+        return grads
+
+    @torch.no_grad()
+    def update_lp_params(self):
+        for i, (bf16_partitions, fp32_partition) in enumerate(zip(self.bf16_partitioned_groups, self.fp32_groups_flat_partition)):
+            partition_id = dist.get_rank(group=self.real_dp_process_group[i])
+            bf16_partitions[partition_id].data.copy_(fp32_partition.data)
+
+    def clear_hp_grads(self):
+        for flat_gradients in self.fp32_groups_gradients_flat:
+            flat_gradients.zero_()
+
+        for group in self.fp32_groups_has_gradients:
+            group = [False] * len(group)
```

Comment:
Are you doing an in-place operation? Don't you have to do something like:
```python
for i in range(len(self.fp32_groups_has_gradients)):
    self.fp32_groups_has_gradients[i] = [False] * len(self.fp32_groups_has_gradients)
```

Reply:
Good catch! Now fixed like this
```
        for i, group in enumerate(self.fp32_groups_gradients):
            self.fp32_groups_has_gradients[i] = [False] * len(group)
```